README.md
# borg-summary

## About

These scripts are intended to run on a backup server hosting multiple [borg backup](http://borgbackup.readthedocs.io/en/stable/index.html) repositories. It can:

* print a succinct summary of a borg repository
* print a summary of all backups in a borg repository
* print a warning if backup data for a borg repository is out of date
* print a warning if there hasn't been a backup in over 24 hours

Because getting backup information from `borg list` can be slow for repositories with many backups, `borgsummary` uses CSV files to store backup information.


## Requirements

* Python 3
* `borgsummary-all` requires `tabulate`


## Usage in cron

A simplified example of using `borgsummary-all` to run hourly checks to update CSV data files, a daily check to ensure backups are running, and a weekly job to send a summary email:

```
@hourly root python3 /root/borg-summary/borgsummary-all.py --update /data/borg
@daily  root python3 /root/borg-summary/borgsummary-all.py --check /data/borg | mail -E -s 'Warning: borg backup issues' root
@weekly root python3 /root/borg-summary/borgsummary-all.py /data/borg | mail -s 'Borg backup summary' root
```


## Borg pool structure and CSV files

To use `borgsummary-all`, the directory hierarchy is expected to be:

```
/some/path/to/backups
    /host1.example.com/borgbackupA
    /host1.example.com/borgbackupB
    /host1.example.com/borgbackupN...
    /host2.example.com/borgbackupA
    ...
```

This "doubled" directory structure is to accommodate clients with multiple borg backup repositories.

Currently, `borgsummary` expects each host to have one backup set, with its name matching the client's hostname.



## Changelog

### [Unreleased]


### [0.1]

#### Added

* Read & write CSV data files containing backup info for each host.
* Better error checking from borg output.
* Add `borgsummary-all` which prints a very succinct summary of all backup repos, and optionally details about every repo.

#### Changed

* Split into two scripts; `borgsummary` reports on just one borg backup; `borgsummary-all` is a wrapper script which handles multiple borg repos.
* Created BorgBackupRepo class to better encapsulate the data in a borg backup repo.

#### Removed

* Because it's no longer applicable, removed `--first` option.

### [Initial version]

* Prints a simple summary of multiple borg backup repositories.


CHANGELOG
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

### Initial version
- Prints a simple summary of multiple borg backup repositories


borg-summary.py
#!/usr/bin/env python3

"""
borg-summary.py

Copyright 2019  John Begenisich

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

import os
from pathlib import Path
import glob
import subprocess
import argparse
import csv
import datetime


BORG_ENV = os.environ.copy()
# TODO: possibly want these configureable
BORG_ENV['BORG_RELOCATED_REPO_ACCESS_IS_OK'] = 'yes'
BORG_ENV['BORG_UNKNOWN_UNENCRYPTED_REPO_ACCESS_IS_OK'] = 'yes'

BACKUP_FIELDS = ['backup_name', 'start_time', 'end_time', 'num_files', 'original_size',
                 'dedup_size', 'all_original_size', 'all_dedup_size', 'command_line']


def get_xdg():
    """
    Return a Path to the XDG_DATA_HOME for borg-summary.
    """
    if 'XDG_DATA_HOME' in os.environ:
        return os.environ['XDG_DATA_HOME'] / '.local' / 'share' / 'borg-summary'
    return Path.home() / '.local' / 'share' / 'borg-summary'


def print_error(message, stdout=None, stderr=None):
    """
    Print an error, optionally include stdout and/or stderr strings, using red for error.
    """
    # TODO: color should be optional
    print(f'\033[0;31m{message}\033[0m')
    if stdout or stderr:
        print('output from borg follows:')
        if stdout:
            print(stdout.decode().strip())
        if stderr:
            print('\033[0;31m{}\033[0m'.format(stderr.decode().strip()))


def get_backup_list(path):
    """
    Return a list of backups in <path> that get be queried with "borg info".
    Returns None if the directory is locked by borgbackup.
    Exits with 1 on error from borg.
    """
    # check for lock file
    if (Path(path) / 'lock.exclusive').exists():
        return None
    result = subprocess.run(['borg', 'list', '--short', str(path)],
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE,
                            env=BORG_ENV)
    if result.returncode != 0:
        print_error('Error running: {}'.format(' '.join(result.args)), result.stdout, result.stderr)
        exit(1)
    backup_list = []
    for line in result.stdout.decode().split('\n'):
        # there is a blank newline at the end
        if line:
            backup_list.append(line)
    return backup_list


def get_backup_info(path, backup_name):
    """
    Returns a dictionary describing borg backup <backup_name> in <path>.
    Exits with 1 on error from borg.
    {
        'start_time': datetime,
        'end_time': datetime,
        'num_files': int,
        'original_size': str,
        'dedup_size': str,
        'all_original_size': str,
        'all_dedup_size': str,
        'command_line': str,
    }
    """
    result = subprocess.run(['borg', 'info', f'{path}::{backup_name}'],
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE,
                            env=BORG_ENV)
    if result.returncode != 0:
        print_error('Error running "{}"'.format(' '.join(result.args)), result.stdout, result.stderr)
        exit(1)
    borg_info = {'backup_name': backup_name}
    lines = result.stdout.decode().split('\n')
    for line in lines:
        s = line.split()
        if line.startswith('Time (start):'):
            borg_info['start_time'] = datetime.datetime.strptime('{} {}'.format(s[3], s[4]), '%Y-%m-%d %H:%M:%S')
        elif line.startswith('Time (end):'):
            borg_info['end_time'] = datetime.datetime.strptime('{} {}'.format(s[3], s[4]), '%Y-%m-%d %H:%M:%S')
        elif line.startswith('Number of files:'):
            borg_info['num_files'] = int(s[3])
        elif line.startswith('This archive:'):
            borg_info['original_size'] = '{} {}'.format(s[2], s[3])
            borg_info['dedup_size'] = '{} {}'.format(s[6], s[7])
        elif line.startswith('All archives:'):
            borg_info['all_original_size'] = '{} {}'.format(s[2], s[3])
            borg_info['all_dedup_size'] = '{} {}'.format(s[6], s[7])
        elif line.startswith('Command line:'):
            borg_info['command_line'] = line[14:]
    return borg_info


def write_backup_data_file(borg_path, data_filename):
    """
    Create a CSV file containing a list of backups for a borg repository,
    overwriting if it already exists.  Skip if borg has locked the repo
    (i.e., a backup is running).
    borg_path: the path to a borg backup pool
    data_filename: the Path to a CSV file describing the backup sets
    """
    backup_names = get_backup_list(borg_path)
    if not backup_names:  # either None - locked by borg; or [] - no backups
        return
    with open(data_filename, 'w', newline='') as csvfile:
        writer = csv.DictWriter(csvfile, BACKUP_FIELDS)
        writer.writeheader()
        for backup_name in backup_names:
            writer.writerow(get_backup_info(borg_path, backup_name))


def read_backup_data_file(data_filename):
    """
    Return a list of dicts representing the backups in a borg repository.
    See get_backup_info() for the dict format.
    data_filename: the Path to a CSV file describing the backup sets
    """
    backup_list = []
    with open(data_filename) as csvfile:
        reader = csv.DictReader(csvfile, BACKUP_FIELDS)
        next(reader, None)  # skip header
        for row in reader:
            row['start_time'] = datetime.datetime.strptime(row['start_time'], '%Y-%m-%d %H:%M:%S')
            row['end_time'] = datetime.datetime.strptime(row['end_time'], '%Y-%m-%d %H:%M:%S')
            backup_list.append(row)
    return backup_list


def get_data_file_age(data_filename):
    """
    Returns an int representing the age of data_filename in number of minutes.
    """
    mtime = datetime.datetime.fromtimestamp(data_filename.stat().st_mtime)
    deltat = datetime.datetime.now() - mtime
    return deltat.days * 1440 + deltat.seconds // 60


def check_data_file_age(backup_name, data_filename):
    """
    Print a warning if Path data_filename is older than 24 hours.
    """
    age_in_days = get_data_file_age(data_filename) // 1440
    if age_in_days >= 1:
        print('Warning: backup {} is {} {} old'.format(backup_name, age_in_days, 'day' if age_in_days == 1 else 'days'))


# -----

def main():
    """
    main
    """
    parser = argparse.ArgumentParser(description='Print a summary of a borgbackup repository',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('path', help='The path to a borgbackup repository')
    parser.add_argument('--csv', type=str, default=None,
                        help='The path to a CSV file holding backup info; generated automatically if not specified.')
    parser.add_argument('--data-path', type=str, default=Path.home() / 'borg-summary',
                        help='The path to CSV data files holding backup info; default: {}'.format(Path.home() / 'borg-summary'))
    parser.add_argument('--update', action='store_true', default=False, help='Create CSV data file')
    parser.add_argument('--autoupdate', action='store_true', default=False,
                        help='Create CSV data file if currente data file is older than 24 hours.')
    parser.add_argument('--check', action='store_true', default=False,
                        help='Print a warning if the CSV data file is older than 24 hours, otherwise no output.')
    args = parser.parse_args()

    borg_path = Path(args.path)
    if not borg_path.is_dir():
        print(f'{borg_path} not found!')
        exit(1)

    if args.csv:
        data_filename = Path(args.csv)
        backup_name = args.csv
    else:
        data_filename = get_xdg() / borg_path.parent.name / (borg_path.name + '.csv')
        backup_name = f'{borg_path.parent.name} - {borg_path.name}'

    if not data_filename.parent.is_dir():
        os.makedirs(data_filename.parent)

    if args.update:
        # TODO: if user specified --update, probably want a warning if couldn't write file (i.e., a backup is running)
        write_backup_data_file(borg_path, data_filename)

    if args.autoupdate:
        if not data_filename.is_file() or get_data_file_age(data_filename) > 1440:
            write_backup_data_file(borg_path, data_filename)

    if not data_filename.is_file():
        print(f'{data_filename} not found!')
        exit(1)

    if args.check:
        check_data_file_age(backup_name, data_filename)
        return

    # normal operation - print summary

    backups = read_backup_data_file(data_filename)

    print(backup_name)
    print('-' * len(backup_name))

    print('\nCommand line: {}\n'.format(backups[-1]['command_line']))

    print('Size of all backups:              {:>10s}'.format(backups[-1]['all_original_size']))
    print('Deduplicated size of all backups: {:>10s}'.format(backups[-1]['all_dedup_size']))
    result = subprocess.check_output('du -sh {}'.format(borg_path), shell=True)
    print('Actual size on disk:              {:>10s}'.format(result.decode().split()[0]))
    print()
    print('{:<16s}  {:<16s}  {:>10s}  {:>10s}  {:>10s}'.format('Start time', 'End time', '# files', 'Orig size', 'Dedup size'))
    print('{:<16s}  {:<16s}  {:>10s}  {:>10s}  {:>10s}'.format('----------', '--------', '-------', '---------', '----------'))
    for d in backups:
        print('{:%Y-%m-%d %H:%M}  {:%Y-%m-%d %H:%M}  {:>10s}  {:>10s}  {:>10s}'.format(d['start_time'], d['end_time'], d['num_files'], d['original_size'], d['dedup_size']))
    print()


if __name__ == '__main__':
    main()


borgsummary-all.py
#!/usr/bin/env python3

"""
borgsummary-all.py

Copyright 2019  John Begenisich

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

import os
from pathlib import Path
import argparse
import subprocess
from tabulate import tabulate
from borgsummary import BorgBackupRepo, size_to_gb


def get_all_repos(pool_path):
    """
    Return a list of Paths in pool_path (a Path), each to a single borg backup repo.
    """
    repos = []
    for host in sorted(os.listdir(pool_path)):
        for repo in sorted(os.listdir(pool_path / host)):
            repos.append(Path(pool_path / host / repo))
    return repos


def get_summary_info_of_all_repos(pool_path):
    """
    Return a list of dicts, each dict containing some data about a borg backup repo.
    all_repos is a list of paths to borg backup repos.
    """
    backup_data = {}
    for borg_path in get_all_repos(pool_path):
        borgbackup = BorgBackupRepo(borg_path)
        backup_data[borgbackup.host] = (borgbackup.repo, borgbackup.read_backup_data_file())
    summaries = []
    for host in backup_data:
        repo, data = backup_data[host]
        if not data:
            # no data!
            summary = {'host': host, 'repo': repo, 'num_backups': 0}
        else:
            last_backup = data[-1]
            duration = last_backup['end_time'] - last_backup['start_time']
            # some info about all backups as well as the most recent one
            summary = {'host': host,
                       'repo': repo,
                       'backup_id': last_backup['backup_id'],
                       'start_time': last_backup['start_time'],
                       'end_time': last_backup['end_time'],
                       'duration': duration,
                       'num_backups': len(data),
                       'num_files': last_backup['num_files'],
                       'original_size': last_backup['original_size'],
                       'dedup_size': last_backup['dedup_size'],
                       'all_original_size': last_backup['all_original_size'],
                       'all_dedup_size': last_backup['all_dedup_size'],
                       'command_line': last_backup['command_line']}
        summaries.append(summary)
    return summaries


def update_all_repos(pool_path):
    for repo_path in get_all_repos(pool_path):
        borg_repo = BorgBackupRepo(repo_path)
        borg_repo.update()


def auto_update_all_repos(pool_path):
    for repo_path in get_all_repos(pool_path):
        borg_repo = BorgBackupRepo(repo_path)
        borg_repo.autoupdate()


def check_all_repos(pool_path):
    for repo_path in get_all_repos(pool_path):
        borg_repo = BorgBackupRepo(repo_path)
        borg_repo.check()


def print_summary_of_all_repos(pool_path):
    # actual size of all backups
    result = subprocess.check_output('du -sBG {}'.format(pool_path), shell=True)
    print('Size of all backups in {}: {} GB\n'.format(pool_path, size_to_gb(result.decode().split()[0])))

    summaries = get_summary_info_of_all_repos(pool_path)

    # first, warn if there are any repos with no backups
    for summary in list(summaries):
        if summary['num_backups'] == 0:
            print('Warning: No backups for {}'.format(summary['backup_id']))
            summaries.remove(summary)

    # check if host == repo for all backups
    # if so, we'll remove the repo field to make the report more succinct
    all_host_eq_repo = True
    for summary in summaries:
        if summary['host'] != summary['repo']:
            all_host_eq_repo = False
            break
    if all_host_eq_repo:
        new_summaries = []
        for summary in list(summaries):
            del summary['repo']
            new_summaries.append(summary)
        summaries = new_summaries

    # two tables, one summarizing the last backup & total backup, and one showing the command line
    tables = [
        ['host', 'repo', 'start_time', 'duration', 'num_files', 'num_backups', 'all_dedup_size'],
        # ['host', 'repo', 'command_line']
    ]

    # prettyify the headers
    replacement_keys = {
        'start_time': 'last backup',
        'num_backups': '# backups',
        'num_files': '# files',
        'all_dedup_size': 'size (GB)',
        'command_line': 'command line',
    }

    for table in tables:
        # only include the fields above
        table_data = []
        for summary in summaries:
            new_summary = {}
            for key in table:
                if key in summary:
                    new_summary[key] = summary[key]
                if key in replacement_keys:
                    new_summary[replacement_keys[key]] = summary[key]
                    del new_summary[key]
            table_data.append(new_summary)
        print(tabulate(table_data, headers='keys'))
        print()

    # print detail about every repo
    for repo_path in get_all_repos(pool_path):
        borgbackup = BorgBackupRepo(repo_path)
        borgbackup.print_summary()
        print()



# -----

def main():
    """
    main
    """
    parser = argparse.ArgumentParser(description='Print a summary of borgbackup repositories',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('pool', help='The root directory of a set of borgbackup repositories')
    # FIXME: use --data-path or remove it
    parser.add_argument('--data-path', type=str, default=Path.home() / 'borg-summary',
                        help='Path to CSV files holding backup info; default: {}'.format(Path.home() / 'borg-summary'))
    parser.add_argument('--update', action='store_true', default=False, help='Create CSV summary file(s)')
    parser.add_argument('--autoupdate', action='store_true', default=False,
                        help='Create CSV data files if current data file is older than 24 hours.')
    parser.add_argument('--check', action='store_true', default=False,
                        help='Print a warning if any CSV file is older than 24 hours.')
    args = parser.parse_args()

    pool_path = Path(args.pool)
    if not os.path.isdir(pool_path):
        print(f'{pool_path} not found!')
        exit(1)

    if args.update:
        update_all_repos(pool_path)
        return

    if args.autoupdate:
        auto_update_all_repos(pool_path)
        return

    if args.check:
        check_all_repos(pool_path)
        return

    print_summary_of_all_repos(pool_path)


if __name__ == '__main__':
    main()


borgsummary.py
#!/usr/bin/env python3

"""
borgsummary.py

Copyright 2019  John Begenisich

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

import os
import subprocess
import argparse
import csv
import datetime
from pathlib import Path
import sqlalchemy
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Integer, Float, String, DateTime, ForeignKey


BORG_ENV = os.environ.copy()
# TODO: possibly want these configureable
BORG_ENV['BORG_RELOCATED_REPO_ACCESS_IS_OK'] = 'yes'
BORG_ENV['BORG_UNKNOWN_UNENCRYPTED_REPO_ACCESS_IS_OK'] = 'yes'

BACKUP_FIELDS = ['backup_id', 'start_time', 'end_time', 'num_files', 'original_size',
                 'dedup_size', 'all_original_size', 'all_dedup_size', 'command_line']

Base = declarative_base()


def get_data_home():
    """
    Return a Path to the XDG_DATA_HOME for borg-summary.
    """
    if 'XDG_DATA_HOME' in os.environ:
        path = Path(os.environ['XDG_DATA_HOME'])
    else:
        path = Path.home() / '.local' / 'share'
    return (path / 'borg-summary.sqlite3').resolve()


def size_to_gb(size):
    """
    Convert string size into GB float.
    size is a string like "43.1 GB" or "37.88 T", etc.
    Must end in 'GB', 'G', 'TB', etc.
    """
    import re
    value = float(re.sub('[^0-9\.]', '', size))  # remove everything except numbers & '.'
    size = size.lower()
    if size.endswith('g') or size.endswith('gb'):
        pass
    elif size.endswith('t') or size.endswith('tb'):
        value = value * 1024
    elif size.endswith('m') or size.endswith('mb'):
        value = value / 1024
    elif size.endswith('k') or size.endswith('kb'):
        value = value / 1024 / 1024
    else:
        # I guess we should throw an exception, but...
        print(f'Warning: size_to_gb() can\'t process "{size}"')
    return round(value, 1)


def print_error(message, stdout=None, stderr=None):
    """
    Print an error, optionally include stdout and/or stderr strings, using red for error.
    """
    # TODO: color should be optional
    print(f'\033[0;31m{message}\033[0m')
    if stdout or stderr:
        print('output from borg follows:')
        if stdout:
            print(stdout.decode().strip())
        if stderr:
            print('\033[0;31m{}\033[0m'.format(stderr.decode().strip()))


def get_repo_id(path):
    """
    Return the borg backup repo ID from the path,
    as gleaned from "borg info".
    """
    # check for lock file
    if (Path(path) / 'lock.exclusive').exists():
        # TODO: throw exception
        return None
    result = subprocess.run(['borg', 'info', str(path)],
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE,
                            env=BORG_ENV)
    if result.returncode != 0:
        print_error('Error running: {}'.format(' '.join(result.args)), result.stdout, result.stderr)
        exit(1)
    for line in result.stdout.decode().split('\n'):
        # there is a blank newline at the end
        if line.startswith('Repository ID:'):
            repo_id = line.split()[-1]
    return repo_id


class BorgBackupRepo(Base):
    __tablename__ = 'repo'
    id = Column(String, primary_key=True)
    location = Column(String)

    def __repr__(self):
        return self.location

    def lock_file_exists(self):
        """
        Check if the lock file for this borg repo exists.
        """
        return (Path(self.location) / 'lock.exclusive').exists()

    def get_backup_list(self):
        """
        Return a list of (id, fingerprint) for this borg back repo that get be queried with "borg info".
        Returns None if the directory is locked by borgbackup.
        Exits with 1 on error from borg.
        """
        # check for lock file
        if self.lock_file_exists():
            return None
        result = subprocess.run(['borg', 'list', '--short', str(self.location)],
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE,
                                env=BORG_ENV)
        if result.returncode != 0:
            print_error('Error running: {}'.format(' '.join(result.args)), result.stdout, result.stderr)
            exit(1)
        backup_list = []
        for line in result.stdout.decode().split('\n'):
            if line:  # there is a blank newline at the end
                backup_list.append(line)
        return backup_list

    def update_backups(self, verbose=False):
        """
        Get list of backups in the borg backup repo, and add any missing
        backups to SQL.
        """
        if self.lock_file_exists():
            if verbose:
                print(f'Cannot update {self.location}; lock file exists')
            return
        backups = self.get_backup_list()
        if not backups:
            # either no backups, or repo is locked (i.e., a backup is running)
            return
        session = Session()
        for backup_id in backups:
            # in SQL?
            backup = session.query(BorgBackup).filter_by(id=backup_id).first()
            if backup is None:
                # this backup does not exist in SQL; add it
                info = self.get_backup_info(backup_id)
                new_backup = BorgBackup(id=info['backup_id'],
                                        repo=self.id,
                                        start_time=info['start_time'],
                                        end_time=info['end_time'],
                                        num_files=info['num_files'],
                                        original_size=info['original_size'],
                                        dedup_size=info['dedup_size'],
                                        all_original_size=info['all_original_size'],
                                        all_dedup_size=info['all_dedup_size'],
                                        command_line=info['command_line']
                                        )
                if verbose:
                    print('adding {}'.format(new_backup))
                session.add(new_backup)
                session.commit()
        session.close()

    def get_backup_info(self, backup_id):
        """
        Returns a dictionary describing the borg backup <backup_id> in our <location>.
        Exits with 1 on error from borg.
        TODO: throw an exception instead of exiting
        """
        result = subprocess.run(['borg', 'info', f'{self.location}::{backup_id}'],
                                stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE,
                                env=BORG_ENV)
        if result.returncode != 0:
            print_error('Error running "{}"'.format(' '.join(result.args)), result.stdout, result.stderr)
            exit(1)
        borg_info = {'backup_id': backup_id}
        lines = result.stdout.decode().split('\n')
        for line in lines:
            s = line.split()
            if line.startswith('Time (start):'):
                borg_info['start_time'] = datetime.datetime.strptime('{} {}'.format(s[3], s[4]), '%Y-%m-%d %H:%M:%S')
            elif line.startswith('Time (end):'):
                borg_info['end_time'] = datetime.datetime.strptime('{} {}'.format(s[3], s[4]), '%Y-%m-%d %H:%M:%S')
            elif line.startswith('Number of files:'):
                borg_info['num_files'] = int(s[3])
            elif line.startswith('This archive:'):
                borg_info['original_size'] = size_to_gb('{} {}'.format(s[2], s[3]))
                borg_info['dedup_size'] = size_to_gb('{} {}'.format(s[6], s[7]))
            elif line.startswith('All archives:'):
                borg_info['all_original_size'] = size_to_gb('{} {}'.format(s[2], s[3]))
                borg_info['all_dedup_size'] = size_to_gb('{} {}'.format(s[6], s[7]))
            elif line.startswith('Command line:'):
                borg_info['command_line'] = line[14:]
        return borg_info

    def print_summary(self):
        """
        Normal operation - print a summary about the backups in this borg backup repo.
        """
        session = Session()
        backups = session.query(BorgBackup).filter_by(repo=self.id).all()
        if not backups:
            print('No backups!')
            session.close()
            return
        print(self.location)
        print('-' * len(self.location))
        print('\nCommand line: {}\n'.format(backups[-1].command_line))
        # TODO: switch to tabulate, I guess
        print('Size of all backups (GB):              {:>8.1f}'.format(backups[-1].all_original_size))
        print('Deduplicated size of all backups (GB): {:>8.1f}'.format(backups[-1].all_dedup_size))
        result = subprocess.check_output('du -sBG {}'.format(self.location), shell=True)
        print('Actual size on disk (GB):              {:>8.1f}'.format(size_to_gb(result.decode().split()[0])))
        print()
        print('{:<16s}  {:<16s}  {:>10s}  {:>10s}  {:>10s}'.format('Start time', 'End time', '# files', 'Orig size', 'Dedup size'))
        print('{:<16s}  {:<16s}  {:>10s}  {:>10s}  {:>10s}'.format('----------', '--------', '-------', '---------', '----------'))
        for backup in backups:
            print('{:%Y-%m-%d %H:%M}  {:%Y-%m-%d %H:%M}  {:>10n}  {:>10.1f}  {:>10.1f}'.format(backup.start_time,
                                                                                        backup.end_time,
                                                                                        backup.num_files,
                                                                                        backup.original_size,
                                                                                        backup.dedup_size))
        session.close()

    def check(self):
        """
        Warn if there haven't been any backups for over 24 hours.
        """
        session = Session()
        backups = session.query(BorgBackup).filter_by(repo=self.id).all()
        if not backups:
            print(f'Warning: no backups for {self.location}')
            session.close()
            return
        # time of backup completion
        last_backup_age_in_days = (datetime.datetime.now() - backups[-1].end_time).days
        if last_backup_age_in_days >= 1:
            print('Warning: {}: no backup for {} {} (last backup finished: '
                  '{:%Y-%m-%d %H:%M})'.format(self.location,
                                              last_backup_age_in_days,
                                              'day' if last_backup_age_in_days == 1 else 'days',
                                              backups[-1].end_time))


class BorgBackup(Base):
    __tablename__ = 'backup'
    id = Column(String, primary_key=True)
    repo = Column(String, ForeignKey('repo.id'), nullable=False)
    start_time = Column(DateTime)
    end_time = Column(DateTime)
    num_files = Column(Integer)
    original_size = Column(Float)
    dedup_size = Column(Float)
    all_original_size = Column(Float)
    all_dedup_size = Column(Float)
    command_line = Column(String)

    def __repr__(self):
        return self.id


# -----

def main():
    """
    main
    """
    parser = argparse.ArgumentParser(description='Print a summary of a borgbackup repository',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('path', help='The path to a borgbackup repository')
    # FIXME: use --data-path or remove it
    # parser.add_argument('--data-path', type=str, default=Path.home() / 'borg-summary',
    #                     help='The path to CSV data files holding backup info; default: {}'.format(Path.home() / 'borg-summary'))
    parser.add_argument('--update', action='store_true', default=False, help='Update SQL from backup repo (if possible)')
    parser.add_argument('--check', action='store_true', default=False,
                        help='Print a warning if no backups in over 24 hours.')
    parser.add_argument('--detail', action='store_true', default=False,
                        help='Print a summary of the backups in this repo.')
    parser.add_argument('-v', '--verbose', action='store_true', default=False, help='Be verbose')
    args = parser.parse_args()

    if not args.detail and not args.update and not args.check:
        print('Must specify at least one of "update", "check", detail"')
        return

    sql_filename = get_data_home()

    global Session
    engine = sqlalchemy.create_engine(f'sqlite:///{sql_filename}', echo=False)
    Base.metadata.create_all(engine)
    Session = sqlalchemy.orm.sessionmaker(bind=engine)
    session = Session()

    location = Path(args.path).resolve()
    repo_id = get_repo_id(location)

    repo = session.query(BorgBackupRepo).filter_by(id=repo_id).first()
    if repo is None:
        # add repo to SQL
        repo = BorgBackupRepo(id=repo_id, location=str(location))
        if args.verbose:
            print('Adding new repo: {}'.format(repo))
        session.add(repo)
        session.commit()

    if args.update:
        repo.update_backups(verbose=args.verbose)

    if args.check:
        repo.check()

    if args.detail:
        repo.print_summary()

    session.close()


if __name__ == '__main__':
    main()
